### PR TITLE
[Translations][Shop][API] Dynamic serialization of translations for a Product, ProductVariant and Taxon entities

### PR DESCRIPTION
**Pull request by Rafikooo**
| Q               | A
| --------------- | -----
| Branch?         | 1.11  
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no 
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Service/SprintfResponseEscaper.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ public static function provideMessageWithEscapedResponseContent(string $message,
             str_replace(
                 '%',
                 '%%',
-                $response->getContent()
+                json_encode(json_decode($response->getContent(), true), JSON_PRETTY_PRINT)
```
Comment (Rafikooo) [inline src/Sylius/Behat/Service/SprintfResponseEscaper.php:28]: I decided to improve JSON output shape
Screenshot 2021-12-15 at 18 40 00
Screenshot 2021-12-15 at 18 39 07
 to make life a little bit easier when debugging

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 2]
```diff
@@ -794,17 +794,40 @@ private function hasItemWithNameAndQuantity(Response $response, string $productN
...
+        $this->checkProductQuantity($cartResponse, $productName, $quantity, 'customer');
+    }
+
+    private function checkProductQuantity(
+        Response $cartResponse,
+        string $productName,
+        int $quantity,
+        string $customerOrAdmin
+    ): void {
+        if (!($customerOrAdmin === 'customer' || $customerOrAdmin === 'admin')) {
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:819]:
```suggestion
        if ($customerOrAdmin !== 'customer' && $customerOrAdmin !== 'admin') {
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductVariant.xml [hunk 3]
```diff
@@ -82,6 +82,7 @@
 
         <property name="id" identifier="false" writable="false" />
         <property name="code" identifier="true" required="true" />
+        <property name="name" identifier="false" required="true" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/ProductVariant.xml:85]:
```suggestion
        <property name="name" required="true" />
```
I guess it is not needed

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Taxon.xml [hunk 4]
```diff
@@ -61,6 +61,8 @@
 
         <property name="id" identifier="false" writable="false" />
         <property name="code" identifier="true" required="true" />
+        <property name="name" identifier="false" required="true" />
+        <property name="description" identifier="false" required="true" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Taxon.xml:65]:
```suggestion
        <property name="name" required="true" />
        <property name="description" required="true" />
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Rafał! :1st_place_medal:

File: tests/Api/Shop/ProductVariantsTest.php [hunk 5]
```diff
@@ -21,22 +22,45 @@ final class ProductVariantsTest extends JsonApiTestCase
...
         $this->client->request('GET', '/api/v2/shop/product-variants/MUG_RED', [], [], self::CONTENT_TYPE_HEADER);
         $response = $this->client->getResponse();
 
-        $this->assertResponse($response, 'shop/get_product_variant_with_price_response', Response::HTTP_OK);
+        $this->assertResponse($response, 'shop/product/get_product_variant_with_price_response', Response::HTTP_OK);
+    }
+
+    /**
+     * @test
+     */
```
Comment (lchrusciel) [inline tests/Api/Shop/ProductVariantsTest.php:46]:
```suggestion
    /** @test */
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 6]
```diff
@@ -794,17 +794,40 @@ private function hasItemWithNameAndQuantity(Response $response, string $productN
...
+
+    private function checkProductQuantityByCustomer(
+        Response $cartResponse,
+        string $productName,
+        int $quantity,
+    ): void {
+        $this->checkProductQuantity($cartResponse, $productName, $quantity, 'customer');
+    }
+
+    private function checkProductQuantity(
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:813]: Personally, I'm not a fan of this solution. In order to reduce duplication, we've introduced 4 if statements

Comment (Rafikooo) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:813]: Honestly, I don't understand what exactly you mean. Could you describe in other words what I should change?

File: tests/Api/Responses/Expected/shop/get_taxonomy_response.json [hunk 7]
```diff
@@ -9,49 +9,17 @@
...
-                    "@id": "\/api\/v2\/shop\/taxon-translations\/@integer@",
-                    "@type": "TaxonTranslation",
-                    "id": "@integer@",
-                    "name": "Mugs",
-                    "slug": "categories\/mugs",
-                    "description": "Some description Lorem ipsum dolor sit amet."
-                }
-            }
+            "name": "Mugs",
+            "description": "Some description Lorem ipsum dolor sit amet."
```
Comment (lchrusciel) [inline tests/Api/Responses/Expected/shop/get_taxonomy_response.json:13]: Lets add slug to the response (or did you had reason not to?). Also, please mention about this change in UPGRADE-API-1.11.md

File: tests/Api/Responses/Expected/shop/product/get_product_with_default_locale_translation.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    "\/api\/v2\/shop\/product-variants\/MUG_BLUE",
+    "\/api\/v2\/shop\/product-variants\/MUG_RED"
+  ],
+  "options": [
+    "\/api\/v2\/shop\/product-options\/COLOR"
+  ],
+  "createdAt": @date@,
+  "updatedAt": @date@,
+  "description": "This is a mug",
+  "shortDescription": "Short mug description",
```
Comment (lchrusciel) [inline tests/Api/Responses/Expected/shop/product/get_product_with_default_locale_translation.json:22]: can you add slug here as well?

**Pull request merged**